pwnlib/fmtstr.py

```diff
@@ -102,7 +108,497 @@ def send_payload(payload):
 
 log = getLogger(__name__)
 
-def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte'):
+SPECIFIER = {
+    1: 'hhn',
+    2: 'hn',
+    4: 'n',
+    8: 'lln',
+}
+
+WRITE_SIZE = {
+    "byte": 1,
+    "short": 2,
+    "int": 4,
+    "long": 8,
+}
+
+def normalize_writes(writes):
+    # make all writes flat
+    writes = { address: flat(data) for address, data in writes.items() }
+
+    # merge adjacent writes (and detect overlaps)
+    merged = []
+    prev_end = -1
+    for address, data in sorted(writes.items(), key=itemgetter(0)):
+        if address < prev_end:
+            raise ValueError("normalize_writes(): data at offset %d overlaps with previous data which ends at offset %d" % (address, prev_end))
+
+        if address == prev_end and merged:
+            merged[-1] = (merged[-1][0], merged[-1][1] + data)
+        else:
+            merged.append((address, data))
+
+        prev_end = address + len(data)
+
+    return merged
+
+# optimization examples (with bytes_written=0)
+#
+# 00 05 00 00     -> %n%5c%n
+# 00 00 05 00 00  -> %n%5c%n
+# 00 00 05 05 00 05  -> need overlapping writes if numbwritten > 5
+
+class AtomWrite(object):
+    """
+    This class represents a write action that can be carried out by a single format string specifier.
+
+    Each write has an address (start), a size and the integer that should be written.
+
+    Additionally writes can have a mask to specify which bits are important.
+    While the write always overwrites all bytes in the range [start, start+size) the mask sometimes allows more
+    efficient execution. For example, assume the current format string counter is at 0xaabb and a write with
+    with integer = 0xaa00 and mask = 0xff00 needs to be executed. In that case, since the lower byte is not covered
+    by the mask, the write can be directly executed with a %hn sequence (so we will write 0xaabb, but that is ok
+    because the mask only requires the upper byte to be correctly written).
+    """
+    __slots__ = ( "start", "size", "integer", "mask" )
+
+    def __init__(self, start, size, integer, mask=None):
+        if mask is None:
+            mask = (1 << (8 * size)) - 1
+        self.start = long(start)
```

Comment:
There is no need to use `long()`, as it is not compatible with python 3, and in python 2 `int()` can actually handle oversize integers, returning long where appropriate.

```suggestion
        self.start = int(start)
```